transformer_engine/jax/cpp_extensions/attention.py

```diff
@@ -1020,6 +897,378 @@ def sharded_impl(
+                return sliced(k), sliced(v)
+
+        return k, v  # fall through
+
+    def pad_kv(self, dk, dv, pad_seq_len):
+        """Pads dk and dv tensors to a sequence length of pad_seq_len."""
+
+        def pad(x, npad):
+            return jnp.pad(x, npad, "constant", constant_values=0.0)
+
+        match self.config.qkv_layout:
+            case NVTE_QKV_Layout.NVTE_BSHD_BS2HD:
+                npad = [[0, 0], [0, pad_seq_len], [0, 0], [0, 0], [0, 0]]
+                return pad(dk, npad), dv
+            case NVTE_QKV_Layout.NVTE_BSHD_BSHD_BSHD:
+                npad = [[0, 0], [0, pad_seq_len], [0, 0], [0, 0]]
+                return pad(dk, npad), pad(dv, npad)
+
+        return dk, dv  # fall through
+
+
+class FusedAttnCPWithAllGatherFwdPrimitive(FusedAttnFwdPrimitive):
+    """
+    Fused Attention Forward with Context Parallelism Primitive
+
+    This context parallel implementation uses all-gather to collect KV inputs from context parallel ranks.
+    """
+
+    @staticmethod
+    def partition(config, mesh, arg_infos, result_infos):
+        # Call base implementation for non-context parallel mesh to avoid unecessary work.
+        is_context_parallel = get_mesh_axis_size(config.cp_axis, mesh) > 1
+        if not is_context_parallel:
+            return FusedAttnFwdPrimitive.partition(config, mesh, arg_infos, result_infos)
+
+        helper = _FusedAttnCPWithAllGatherHelper(mesh, config)
+        helper.check_supported()
+
+        out_sharding = result_infos[0].sharding
+        softmax_aux_sharding = result_infos[1].sharding
+        rng_state_sharding = seed_sharding = NamedSharding(
+            mesh, PartitionSpec(get_all_mesh_axes(), None)
+        )
+        arg_shardings = tuple([arg_i.sharding for arg_i in arg_infos[:-1]] + [seed_sharding])
+        out_shardings = (out_sharding, softmax_aux_sharding, rng_state_sharding)
+
+        def impl(q, k, v, bias, q_seqlen, kv_seqlen, q_seq_offsets, k_seq_offsets, seed):
+
+            cp_size = get_mesh_axis_size(config.cp_axis, mesh)
+            cp_rank = get_mesh_axis_rank(config.cp_axis, mesh)
+
+            # cuDNN does not support right-aligned masking with dynamic sequence length padding.
+            # Therefore we must explicitly instantiate each CP rank slicing and use a runtime switch
+            # to select the appropriate computation. Each case generates a [..., SEQ/CP, ..] tensor
+            # meeting the expectation of the SPMD model.
+            # TODO(mgoldfarb-nvidia): When cuDNN supports we should be able to make use of a padding
+            # mask/sequence length tensor to avoid this unrolled loop.
+            def _cross_attn(idx, q, k, v, bias, q_seqlen, kv_seqlen, seed):
+                kv_max_seqlen = k.shape[1]
+                kv_seqlen_per_subrank = kv_max_seqlen // (cp_size * 2)
+                assert kv_max_seqlen % cp_size == 0, "sequence length must evenly divide cp size"
+
+                q_split = jnp.split(q, 2, axis=1)
```

Comment:
I think the reason for spliting q here is for load balancing, right? When load balancing is enabled, is it the user's responsibility to reorder the q sequence?

Reply:
Yes this is part of load balance the cp ranks. We split the entire set of sequences into `CP*2` groups. The user needs to call `reorder_causal_load_balancing` prior to attention or ideally at the beginning of a transformer layer stack.

<img width="377" alt="Screenshot 2024-08-15 at 10 38 39 AM" src="https://github.com/user-attachments/assets/47270fe2-4671-4dd7-a3df-f15c77c1508f">